tao_compiler/mlir/disc/transforms/disc_lhlo_rewriter.cc

```diff
@@ -0,0 +1,173 @@
+
+// This file implements logic for lowering HLO DISC dialect to LHLO DISC
+// dialect.
+
+#include <algorithm>
+#include <utility>
+
+#include "llvm/Support/Debug.h"
+#include "mlir-hlo/Dialect/lhlo/IR/lhlo_ops.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Bufferization/IR/Bufferization.h"
+#include "mlir/Dialect/Bufferization/Transforms/Bufferize.h"
+#include "mlir/Dialect/Func/Transforms/FuncConversions.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/Shape/IR/Shape.h"
+#include "mlir/Dialect/Shape/Transforms/Passes.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/AffineMap.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Location.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"  // from @llvm-project
+#include "tensorflow/compiler/mlir/disc/IR/disc_ral_ops.h"
+#include "tensorflow/compiler/mlir/disc/IR/disc_shape_ops.h"
+#include "tensorflow/compiler/mlir/disc/IR/hlo_disc_ops.h"
+#include "tensorflow/compiler/mlir/disc/IR/lhlo_disc_ops.h"
+#include "tensorflow/compiler/mlir/disc/transforms/PassDetail.h"
+#include "tensorflow/compiler/mlir/disc/transforms/disc_map_hlo_to_lhlo_op.h"
+#include "tensorflow/compiler/mlir/disc/transforms/placement_utils.h"
+#include "tensorflow/compiler/mlir/disc/transforms/rewriters.h"
+#include "tensorflow/compiler/mlir/disc/transforms/shape_utils.h"
+
+namespace mlir {
+using placement_utils::kDiscPlaceAssignment;
+using placement_utils::kGpu;
+
+namespace mhlo_disc {
+namespace {
+
+template <typename T>
+using BaseOpConversion = OpConversionPattern<T>;
+
+struct HloToLhloConcatenateOpConverter
+    : public OpRewritePattern<lmhlo::ConcatenateOp> {
+  explicit HloToLhloConcatenateOpConverter(MLIRContext* context)
+      : OpRewritePattern(context) {}
+
+  bool isFixedShape(lmhlo::ConcatenateOp op) const {
+    bool one_size_input = true;
+    bool is_static_shape = true;
+    int num_input_operands = op.getNumOperands() - 1;
+    size_t axis = op.getDimension();
+    disc_ral::ShapeConstraintIRAnalysis shape_analysis(op.getOperation());
+    for (int i = 0; i < num_input_operands; ++i) {
+      auto typ = op.getOperand(i).getType().cast<MemRefType>();
+      if (is_static_shape && typ.hasStaticShape()) {
+        if (one_size_input && i > 0) {
+          auto prev_typ = op.getOperand(i - 1).getType().cast<MemRefType>();
+          one_size_input = (typ.getShape()[axis] == prev_typ.getShape()[axis]);
+        }
+      } else {
+        is_static_shape = false;
+        if (one_size_input && i > 0) {
+          one_size_input = shape_analysis.isShapeEqual(op->getOperand(i),
```

Comment:
Only check concatenating dimension equality for static shape operands while whole shape equality is required for dynamic shape? why're they different?

Reply:
I make this function simpler and clearly because concatent's input shape are equal except for the axis dim, so we only need to check all the operands' shape are equal.